abaco.go

```diff
@@ -388,17 +406,171 @@ func enumerateAbacoRings() (rings []int, err error) {
 
 //------------------------------------------------------------------------------------------------
 
-// AbacoSource represents all AbacoRing ring buffers that can potentially supply data, as well
-// as all AbacoGroups that are discovered in the SampleData phase.
+// AbacoUDPReceiver represents a single Abaco device producing data by UDP packets to a single UDP port.
+type AbacoUDPReceiver struct {
+	host     string       // in the form: "127.0.0.1:56789"
+	conn     *net.UDPConn // active UDP connection
+	data     chan []*packets.Packet
+	sendmore chan bool
+}
+
+// NewAbacoUDPReceiver creates a new AbacoUDPReceiver and binds as a server to the requested host:port
+func NewAbacoUDPReceiver(hostport string) (dev *AbacoUDPReceiver, err error) {
+	dev = new(AbacoUDPReceiver)
+	dev.host = hostport
+	if _, err := net.ResolveUDPAddr("udp", hostport); err != nil {
+		return nil, err
+	}
+	return dev, nil
+}
+
+// start starts the UDP device by...?
+func (device *AbacoUDPReceiver) start() (err error) {
+	raddr, err := net.ResolveUDPAddr("udp", device.host)
+	if err != nil {
+		return err
+	}
+	conn, err := net.ListenUDP("udp", raddr)
+	if err != nil {
+		return err
+	}
+	const UDPPacketSize = 8192
+	bufferPool := sync.Pool{
+		New: func() interface{} { return make([]byte, UDPPacketSize) },
+	}
+	device.conn = conn
+
+	device.sendmore = make(chan bool)
+	device.data = make(chan []*packets.Packet)
+	singlepackets := make(chan []byte, 20)
+	// This goroutine handles UDP message sent one at a time on singlepackets by the other goroutine.
+	// It converts them into packet.Packet objects, queues them into a slice of *packets.Packet
+	// pointers and sends the whole slice when a request comes on device.sendmore.
+	go func() {
+		defer close(device.data)
+		defer func() { device.conn = nil }()
+
+		const initialQueueCapacity = 20000
+		queue := make([]*packets.Packet, 0, initialQueueCapacity)
+		for {
+			select {
+			case _, ok := <-device.sendmore:
+				device.data <- queue
+				if !ok {
+					return
+				}
+				queue = make([]*packets.Packet, 0, initialQueueCapacity)
+
+			case message := <-singlepackets:
+				p, err := packets.ReadPacket(bytes.NewReader(message))
+				bufferPool.Put(message)
+				if err != nil {
+					if err != io.EOF {
+						fmt.Printf("Error converting UDP to packet: err %v, packet %v\n", err, p)
+					}
+					return
+				}
+				queue = append(queue, p)
+			}
+		}
+	}()
+	// This goroutine reads the UDP socket and puts the resulting packets on channel singlepackets.
+	// They will be removed and queued by the previous goroutine.
+	go func() {
+		defer close(singlepackets)
+		for {
+			message := bufferPool.Get().([]byte)
+			if _, _, err := device.conn.ReadFrom(message); err != nil {
+				// Don't report errors. Getting an error here is the normal way to detect closed connection.
```

Comment:
shouldn't we respond to a closed connection?Like by returning?

Reply:
Ohh I see we do return.... i think the comment could be more useful about what we do do instead of about what not to do.

Reply:
Sure. The comment was about the code that's now commented out. I'll clarify.